lib/ace/client/catalog.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+require 'net/http'
+require 'ace/error'
+
+module ACE
+  module Client
+    class Catalog
+      def initialize(config)
+        unless config.is_a? ACE::Config
+          raise ACE::Error.new('`config` must be an ACE::Config',
+                               'puppetlabs/ace/invalid_param')
+        end
+        @config = config
+      end
+
+      def retrieve(certname, environment, facts, trusted, transaction_uuid, job_id)
+        raise ACE::Error.new('`facts` must be a Hash', 'puppetlabs/ace/invalid_param') unless facts.is_a? Hash
```

Comment:
Q - do we want to return nil (or some other 'we have errored' value) if ACE::Error is raised, or are we happy enough that an Error can be caught

Reply:
Good catch, this was bad. I've changed code further down to raise the errors rather than return them to keep consistency.